arelle/ModelXbrl.py

```diff
@@ -266,73 +191,82 @@ class ModelXbrl:
+        self.qnameDimensionDefaults: dict[str, str] = {}  # contains qname of dimension (index) and default member(value)
+        self.facts: list[ModelInlineFact] = []
+        self.factsInInstance: set[ModelInlineFact] = set()
+        self.undefinedFacts: list[ModelFact] = []  # elements presumed to be facts but not defined
+        self.contexts: dict[str, modeldoc.xmlRootElement] = {}
+        self.units: dict[str, ModelUnit] = {}
+        self.modelObjects: list[ModelObject] = []
+        self.qnameParameters: dict[QName, Any] = {}
+        self.modelVariableSets: set[Any] = set()
+        self.modelConsistencyAssertions: set[Any] = set()
+        self.modelCustomFunctionSignatures: dict[QName, Any] = {}
+        self.modelCustomFunctionImplementations: set[modeldoc] = set()
+        self.modelRenderingTables: Set[Any] = set()
         if not keepViews:
-            self.views = []
-        self.langs = {self.modelManager.defaultLang}
+            self.views: list[Any] = []
+        self.langs: set[str] = {self.modelManager.defaultLang}
         from arelle.XbrlConst import standardLabel
-        self.labelroles = {standardLabel}
-        self.hasXDT = False
-        self.hasTableRendering = False
-        self.hasTableIndexing = False
-        self.hasFormulae = False
+        self.labelroles: set[str] = {standardLabel}
+        self.hasXDT: bool = False
+        self.hasTableRendering: bool = False
+        self.hasTableIndexing: bool = False
+        self.hasFormulae: bool = False
         self.formulaOutputInstance = None
-        self.logger = logging.getLogger("arelle")
-        self.logRefObjectProperties = getattr(self.logger, "logRefObjectProperties", False)
-        self.logRefHasPluginAttrs = any(True for m in pluginClassMethods("Logging.Ref.Attributes"))
-        self.logRefHasPluginProperties = any(True for m in pluginClassMethods("Logging.Ref.Properties"))
-        self.logRefFileRelUris = defaultdict(dict)
-        self.profileStats = {}
-        self.schemaDocsToValidate = set()
-        self.modelXbrl = self # for consistency in addressing modelXbrl
-        self.arelleUnitTests = {} # unit test entries (usually from processing instructions
+        self.logger: Any = logging.getLogger("arelle")
+        self.logRefObjectProperties: Any = getattr(self.logger, "logRefObjectProperties", False)
+        self.logRefHasPluginAttrs: bool = any(True for m in pluginClassMethods("Logging.Ref.Attributes"))
+        self.logRefHasPluginProperties: bool = any(True for m in pluginClassMethods("Logging.Ref.Properties"))
+        self.logRefFileRelUris: DefaultDict[Any, dict[str, str]] = defaultdict(dict)
+        self.profileStats: dict[str, tuple[int, Any, Any]] = {}
+        self.schemaDocsToValidate: set[modeldoc] = set()
+        self.closeFileSource: bool
+        self.fileSource: Any
+        self.entryLoadingUrl: Any
+        self.ixdsDocUrls: list[str]
```

Comment:
These uninitialized class field types should be defined up at the class level (where the existing `contexts` and `dimensionDefaultConcepts` fields are)